File: android_app/lib/core/services/logger_service.dart
// ─────────────────────────────────────────────
// PocketTX – Logger Service
// Structured logging with in-memory ring buffer (200 entries)
// + optional file sink. Matches Windows Companion log format.
// ─────────────────────────────────────────────

import 'dart:async';
import 'dart:collection';
import 'package:logger/logger.dart' as pkg_logger;
import '../../models/log_entry_model.dart';
import '../constants/app_constants.dart';

/// Structured logging service with a 200-entry in-memory ring buffer.
class LoggerService {
  static final LoggerService _instance = LoggerService._internal();
  factory LoggerService() => _instance;
  LoggerService._internal();

  final Queue<LogEntry> _buffer = Queue();
  int _nextId = 0;

  final _streamController =
      StreamController<LogEntry>.broadcast();

  final _pkg = pkg_logger.Logger(
    printer: pkg_logger.PrettyPrinter(
      methodCount: 0,
      errorMethodCount: 5,
      lineLength: 80,
      colors: true,
      printEmojis: true,
    ),
  );

  /// Stream of new log entries for reactive UI.
  Stream<LogEntry> get stream => _streamController.stream;

  /// Read-only snapshot of the current ring buffer.
  List<LogEntry> get entries => List.unmodifiable(_buffer.toList());

  int get entryCount => _buffer.length;

  void info(
    LogCategory category,
    String event,
    String message, {
    Map<String, dynamic>? metadata,
  }) =>
      _log(LogLevel.info, category, event, message, metadata: metadata);

  void warning(
    LogCategory category,
    String event,
    String message, {
    Map<String, dynamic>? metadata,
  }) =>
      _log(LogLevel.warning, category, event, message, metadata: metadata);

  void error(
    LogCategory category,
    String event,
    String message, {
    Map<String, dynamic>? metadata,
  }) =>
      _log(LogLevel.error, category, event, message, metadata: metadata);

  void debug(
    LogCategory category,
    String event,
    String message, {
    Map<String, dynamic>? metadata,
  }) =>
      _log(LogLevel.debug, category, event, message, metadata: metadata);

  void verbose(
    LogCategory category,
    String event,
    String message, {
    Map<String, dynamic>? metadata,
  }) =>
      _log(LogLevel.verbose, category, event, message, metadata: metadata);

  void _log(
    LogLevel level,
    LogCategory category,
    String event,
    String message, {
    Map<String, dynamic>? metadata,
  }) {
    final entry = LogEntry(
      id: _nextId++,
      timestamp: DateTime.now(),
      category: category,
      event: event,
      level: level,
      message: message,
      metadata: metadata,
    );

    // Ring buffer: evict oldest when full
    if (_buffer.length >= AppConstants.maxLogEntries) {
      _buffer.removeFirst();
    }
    _buffer.addLast(entry);

    if (!_streamController.isClosed) {
      _streamController.add(entry);
    }

    // Console output
    switch (level) {
      case LogLevel.verbose:
      case LogLevel.debug:
        _pkg.d('[${category.name}] $event: $message');
      case LogLevel.info:
        _pkg.i('[${category.name}] $event: $message');
      case LogLevel.warning:
        _pkg.w('[${category.name}] $event: $message');
      case LogLevel.error:
        _pkg.e('[${category.name}] $event: $message');
    }
  }

  void clear() {
    _buffer.clear();
    _nextId = 0;
  }

  void dispose() {
    _streamController.close();
  }
}

File: companion/src/PocketTX.Companion.Logging/Engine/LoggerService.cs
using PocketTX.Companion.Core.Contracts;
using PocketTX.Companion.Core.Enums;
using PocketTX.Companion.Logging.Models;
using PocketTX.Companion.Logging.Sinks;

namespace PocketTX.Companion.Logging.Engine;

public sealed class LoggerService : ILoggerService
{
    private readonly IEnumerable<ILogSink> _sinks;

    public LoggerService(IEnumerable<ILogSink> sinks)
    {
        _sinks = sinks;
    }

    public void LogDebug(string message, string category = "General", Dictionary<string, object>? metadata = null)
        => Log(LogLevel.Debug, message, category, null, metadata);

    public void LogInfo(string message, string category = "General", Dictionary<string, object>? metadata = null)
        => Log(LogLevel.Info, message, category, null, metadata);

    public void LogWarning(string message, string category = "General", Dictionary<string, object>? metadata = null)
        => Log(LogLevel.Warning, message, category, null, metadata);

    public void LogError(string message, Exception? exception = null, string category = "General", Dictionary<string, object>? metadata = null)
        => Log(LogLevel.Error, message, category, exception, metadata);

    private void Log(LogLevel level, string message, string category, Exception? exception, Dictionary<string, object>? metadata)
    {
        LogEntry entry = new(DateTime.UtcNow, level, category, message, exception, metadata);
        foreach (var sink in _sinks)
        {
            _ = sink.EmitAsync(entry);
        }
    }
}
